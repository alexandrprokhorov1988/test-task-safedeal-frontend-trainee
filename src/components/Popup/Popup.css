.popup {
  margin: 0 auto;
  position: fixed;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  z-index: 1;
  background-color: rgba(0, 0, 0, 0.7);
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  visibility: hidden;
  opacity: 0;
  transition: all ease 0s;
}

.popup_opened {
  transition: all ease .3s;
  visibility: visible;
  opacity: 1;
}

.popup__container {
  padding: 30px;
  box-sizing: border-box;
  background-color: #FFFFFF;
  display: grid;
  max-width: 619px;
  width: 100%;
  height: 425px;
  grid-template-columns: minmax(190px, 331px) minmax(140px, 214px);
  grid-template-rows: 1fr 1fr;
  grid-template-areas: "img comments-container" "form comments-container";
  position: relative;
  justify-items: center;
  grid-column-gap: 14px;
}

.popup__close-button {
  background-color: transparent;
  background-image: url(../../images/popup-close-button.svg);
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  position: absolute;
  right: 21px;
  top: 21px;
  width: 18px;
  height: 18px;
  border: none;
  transition: all ease .1s;
  cursor: pointer;
  transform: rotate(90deg);
}

.popup__close-button:hover {
  transform: rotate(180deg);
}

.popup__img {
  grid-area: img;
  width: 100%;
  object-fit: cover;
  height: 205px;
}

.popup__comments-container {
  grid-area: comments-container;
  width: 100%;
  height: 100%;
  overflow: auto;
  direction: rtl;
}

.popup__comments-container::-webkit-scrollbar {
  width: 16px;
}

.popup__comments-container::-webkit-scrollbar-thumb {
  border: 1px solid #CCCCCC;
}

.popup__comments-container::-webkit-scrollbar-thumb:hover {
  border: 1px solid #878787;
}

.popup__comment-container {
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  padding-left: 5px;
}

.popup__comment-container:not(:first-of-type) {
  padding-top: 20px;
}

.popup__comment {
  padding: 0;
  margin: 0;
  font-size: 13px;
  line-height: 15px;
  color: #000000;
  padding-top: 5px;
}

.popup__comment_type_date {
  color: #999999;
}

.popup__form {
  grid-area: form;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-self: end;
}

.popup__form-error {
  visibility: visible;
  color: #ff2b06;
  font-size: 9px;
  height: 20px;
  line-height: 20px;
}

.popup__form-error_hide {
  visibility: hidden;
}

.popup__form-input {
  font-family: inherit;
  border-radius: 3px;
  border: 1px solid #CCCCCC;
  height: 30px;
  box-sizing: border-box;
  width: 100%;
}

.popup__form-button {
  height: 30px;
  background: #0a6bd0d9;
  border-radius: 3px;
  color: #FFFFFF;
  font-size: 13px;
  line-height: 15px;
  border: none;
  cursor: pointer;
  margin-top: 19px;
  /*transition: all ease .5s;*/
}

.popup__form-button:hover {
  background: #0a54a8;
}

.popup__form-button:disabled {
  background: #4997D0;
  cursor: none;
  pointer-events: none;
}

@media screen and (max-width: 425px) {
  .popup__container {
    height: 100%;
    grid-template-columns: 1fr;
    grid-template-rows: auto 1fr auto;
    grid-template-areas: "img" "comments-container" "form";
    grid-column-gap: 0;
    padding: 0 0 40px 0;
    grid-row-gap: 20px;
  }

  .popup__form,
  .popup__comments-container {
    width: 90%;
  }

  .popup__comments-container {
    height: 100%;
    direction: ltr;
  }

  .popup__comment-container {
    align-items: flex-start;
  }

  .popup__comment {
    padding-top: 0;
    padding-left: 0;
  }
}
