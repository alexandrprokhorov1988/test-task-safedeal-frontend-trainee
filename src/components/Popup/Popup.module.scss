$font-size: 13px;
$line-height: 15px;

.popup {
  margin: 0 auto;
  position: fixed;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  z-index: 1;
  background-color: rgba(0, 0, 0, 0.7);
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  visibility: hidden;
  opacity: 0;
  transition: all ease 0s;
}

.opened {
  transition: all ease .3s;
  visibility: visible;
  opacity: 1;
}

.container {
  padding: 30px;
  box-sizing: border-box;
  background-color: #FFFFFF;
  display: grid;
  max-width: 619px;
  width: 100%;
  height: 425px;
  grid-template-columns: minmax(190px, 331px) minmax(140px, 214px);
  grid-template-rows: 1fr 1fr;
  grid-template-areas: "img comments-container" "form comments-container";
  position: relative;
  justify-items: center;
  grid-column-gap: 14px;
}

.closeButton {
  background-color: transparent;
  background-image: url(assets/popup-close-button.svg);
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  position: absolute;
  right: 21px;
  top: 21px;
  width: 18px;
  height: 18px;
  border: none;
  transition: all ease .1s;
  cursor: pointer;
  transform: rotate(90deg);

  &:hover {
    transform: rotate(180deg);
  }
}

.img {
  grid-area: img;
  width: 100%;
  object-fit: cover;
  height: 205px;
}

.commentsContainer {
  grid-area: comments-container;
  width: 100%;
  height: 100%;
  overflow: auto;
  direction: rtl;
}

.commentsContainer::-webkit-scrollbar {
  width: 16px;
}

.commentsContainer::-webkit-scrollbar-thumb {
  border: 1px solid #CCCCCC;
}

.commentsContainer::-webkit-scrollbar-thumb:hover {
  border: 1px solid #878787;
}

.commentContainer {
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  padding-left: 5px;
}

.commentContainer:not(:first-of-type) {
  padding-top: 20px;
}

.comment {
  font-size: $font-size;
  line-height: $line-height;
  padding: 0;
  margin: 0;
  color: #000000;
  padding-top: 5px;
}

.commentTypeDate {
  color: #999999;
}

.form {
  grid-area: form;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-self: end;
}

.formError {
  visibility: visible;
  color: #ff2b06;
  font-size: 9px;
  height: $font-size + 7;
  line-height: $line-height + 5;
}

.formErrorHide {
  visibility: hidden;
}

.formInput {
  font-family: inherit;
  border-radius: 3px;
  border: 1px solid #CCCCCC;
  height: 30px;
  box-sizing: border-box;
  width: 100%;
}

.formButton {
  height: 30px;
  background: #0a6bd0d9;
  border-radius: 3px;
  color: #FFFFFF;
  font-size: $font-size;
  line-height: $line-height;
  border: none;
  cursor: pointer;
  margin-top: 19px;

  &:hover {
    background: #0a54a8;
  }

  &:disabled {
    background: #4997D0;
    cursor: none;
    pointer-events: none;
  }
}

@media screen and (max-width: 425px) {
  .container {
    height: 100%;
    grid-template-columns: 1fr;
    grid-template-rows: auto 1fr auto;
    grid-template-areas: "img" "comments-container" "form";
    grid-column-gap: 0;
    padding: 0 0 40px 0;
    grid-row-gap: 20px;
  }

  .form,
  .commentsContainer {
    width: 90%;
  }

  .commentsContainer {
    height: 100%;
    direction: ltr;
  }

  .commentContainer {
    align-items: flex-start;
  }

  .popupComment {
    padding-top: 0;
    padding-left: 0;
  }
}
