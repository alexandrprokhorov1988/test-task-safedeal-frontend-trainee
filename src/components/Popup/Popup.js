import React from 'react';
import './Popup.css';
import {useFormValidation} from "../../hooks/useFormValidation";

const Popup = React.memo((
  {
    isPopupOpen,
    onClose,
    onOpen,
    onSubmit,
    isLoadingComment,
    onSubmitComment,
    currentOriginSizeImage,
    comments
  }) => {
  const { values, errors, isValid, handleChange, resetForm } = useFormValidation();

  function handleSubmit(e) {
    e.preventDefault();
    if (values.name === '' || values.comment === '') {
      return;
    }
    onSubmitComment(values, currentOriginSizeImage.id);
    resetForm();
  }

  return (
    <div className={`popup ${isPopupOpen && "popup_opened"}`} onClick={onClose}>
      <div className="popup__container" onClick={(e) => e.stopPropagation()}>
        <button
          className="popup__close-button"
          type="button"
          onClick={onClose}
        />
        <img className="popup__img" src={currentOriginSizeImage.url} alt="Картинка"/>
        <div className="popup__comments-container">
          {comments.map((comment, index) => (
            <div key={comment.id} className="popup__comment-container">
              <p className="popup__comment popup__comment_type_date">{comment.date}</p>
              <p className="popup__comment">{comment.text}</p>
            </div>
          ))}
        </div>
        <form className="popup__form" action="#" noValidate method="get" onSubmit={handleSubmit}>
          <span
            className={`popup__form-error ${isValid && 'popup__form-error_hide'}`}
          >{errors.name || ''}</span>
          <input
            name="name"
            className="popup__form-input"
            type="text"
            placeholder="Ваше имя"
            value={values.name || ''}
            onChange={handleChange}
            required
            minLength="1"
            maxLength="200"
            pattern="^[а-яёА-ЯЁa-zA-Z0-9-\s\-]+$"
          />
          <span
            className={`popup__form-error ${isValid && 'popup__form-error_hide'}`}
          >{errors.comment || ''}</span>
          <input
            name="comment"
            className="popup__form-input"
            type="text"
            placeholder="Ваш комментарий"
            value={values.comment || ''}
            onChange={handleChange}
            required
            minLength="2"
            maxLength="200"
            pattern="^[а-яёА-ЯЁa-zA-Z0-9-\s\-]+$"
          />
          <button
            className="popup__form-button"
            type="submit"
            disabled={isLoadingComment || !isValid}
          >{`${isLoadingComment ? 'Добавление комментария' : 'Оставить комментарий'}`}
          </button>
        </form>
      </div>
    </div>
  );
});

export default Popup;
