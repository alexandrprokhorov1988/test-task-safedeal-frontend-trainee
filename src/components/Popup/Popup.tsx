import React from 'react';
import './Popup.css';
import {useFormValidation} from "../../hooks/useFormValidation";
import mainApi from "../../utils/mainApi";
import {dateParseFromTimestampToString} from "../../utils/helpers";
import {useAppSelector} from "../../hooks/useAppSelector";
import {useAppDispatch} from "../../hooks/useAppDispatch";

interface IPopupProps {
  onClose: () => void;
}

const Popup: React.FC<IPopupProps> = ({onClose}) => {

  const {values, errors, isValid, handleChange, resetForm} = useFormValidation();
  const main = useAppSelector(state => state.main);
  const popup = useAppSelector(state => state.popup);
  const photoGrid = useAppSelector(state => state.photoGrid);
  const {setIsLoadingComment, setComments} = useAppDispatch();

  console.log('popup');  //todo del

  React.useEffect(() => {
    resetForm();
  }, [photoGrid.comments, resetForm]);

  function handleSubmitComment( name: string, comment: string , id: number) {
    setIsLoadingComment(true);
    return mainApi.setNewComment({
      id: id,
      name: name,
      comment: comment
    })
      .then(() => {
        const answerFromServer = {
          id: Math.random(),
          text: values.comment,
          date: dateParseFromTimestampToString(new Date().getTime())
        };
        setComments([...photoGrid.comments, answerFromServer]);
      })
      .catch((err) => console.log(err))
      .finally(() => setIsLoadingComment(false));
  }

  function handleSubmit(e: React.FormEvent<HTMLFormElement>): void {
    e.preventDefault();
    if (values.name === '' || values.comment === '') {
      return;
    }
    handleSubmitComment(values.name!, values.comment!, photoGrid.currentOriginSizeImage.id!);
  }

  return (
    <div className={`popup ${main.isOpenPopup ? "popup_opened" : ""}`} onClick={onClose} role="button" tabIndex={-1}
         aria-hidden="true">
      <div className="popup__container" onClick={(e) => e.stopPropagation()} role="complementary" tabIndex={-1}
           aria-hidden="true">
        <button
          className="popup__close-button"
          type="button"
          onClick={onClose}
        />
        <img className="popup__img" src={photoGrid.currentOriginSizeImage.url} alt="Картинка"/>
        <div className="popup__comments-container">
          {photoGrid.comments.map((comment) => (
            <div key={comment.id} className="popup__comment-container">
              <p className="popup__comment popup__comment_type_date">{comment.date}</p>
              <p className="popup__comment">{comment.text}</p>
            </div>
          ))}
        </div>
        <form className="popup__form" action="#" noValidate method="get" onSubmit={handleSubmit}>
          <span
            className={`popup__form-error ${!errors.name && 'popup__form-error_hide'}`}
          >{errors.name || ''}</span>
          <input
            name="name"
            className="popup__form-input"
            type="text"
            placeholder="Ваше имя"
            value={values.name || ''}
            onChange={handleChange}
            required
            minLength={2}
            maxLength={200}
            pattern="^[а-яёА-ЯЁa-zA-Z0-9-\s\-]+$"
          />
          <span
            className={`popup__form-error ${!errors.comment && 'popup__form-error_hide'}`}
          >{errors.comment || ''}</span>
          <input
            name="comment"
            className="popup__form-input"
            type="text"
            placeholder="Ваш комментарий"
            value={values.comment || ''}
            onChange={handleChange}
            required
            minLength={2}
            maxLength={200}
            pattern="^[а-яёА-ЯЁa-zA-Z0-9-\s\-]+$"
          />
          <button
            className="popup__form-button"
            type="submit"
            disabled={popup.isLoadingComment || !isValid}
          >{`${popup.isLoadingComment ? 'Добавление комментария' : 'Оставить комментарий'}`}
          </button>
        </form>
      </div>
    </div>
  );
};

export default React.memo(Popup);

