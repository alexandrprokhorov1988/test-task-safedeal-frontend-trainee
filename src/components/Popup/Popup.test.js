import React from 'react';
import {render, unmountComponentAtNode} from "react-dom";
import renderer from 'react-test-renderer';
import {act} from "react-dom/test-utils";
import Popup from './Popup';
import closePopup from '../../components/App/App';
import handleSubmitComment from '../../components/App/App';

let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});
afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});
describe('Popup component', () => {
  it('Должна вызывать коллбэк функцию по click на кнопку закрытия попапа', () => {
    const onClose = jest.fn();
    act(() => {
      render(<Popup
        isOpenPopup={true}
        onClose={onClose}
        isLoadingComment={false}
        onSubmitComment={onClose}
        comments={[{ text: "Крутая фотка", date: "03.01.2020", id: 153 }]}
        currentOriginSizeImage={{ id: 237, url: "https://picsum.photos/id/237/600/400" }}
      />, container);
    });
    const buttonClose = document.querySelector(".popup__close-button");
    act(() => {
      buttonClose.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('Должна вызывать коллбэк функцию по click на кнопку submit', () => {
    const onSubmitComment = jest.fn();
    act(() => {
      render(<Popup
        isOpenPopup={true}
        onClose={onSubmitComment}
        isLoadingComment={false}
        onSubmitComment={onSubmitComment}
        comments={[{ text: "Крутая фотка", date: "03.01.2020", id: 153 }]}
        currentOriginSizeImage={{ id: 237, url: "https://picsum.photos/id/237/600/400" }}
      />, container);
    });
    const buttonSubmit = document.querySelector(".popup__form-button");

    act(() => {
      buttonSubmit.dispatchEvent(new Event("submit", { bubbles: true }));
    });
    expect(onSubmitComment).toHaveBeenCalledTimes(1);
  });
  
  it('Popup snapshot', () => {
    const render = renderer
      .create(<Popup
        isOpenPopup={true}
        onClose={closePopup}
        isLoadingComment={false}
        onSubmitComment={handleSubmitComment}
        comments={[{ text: "Крутая фотка", date: "03.01.2020", id: 153 }]}
        currentOriginSizeImage={{ id: 237, url: "https://picsum.photos/id/237/600/400" }}
      />)
      .toJSON();
    expect(render).toMatchSnapshot();
  });
});
