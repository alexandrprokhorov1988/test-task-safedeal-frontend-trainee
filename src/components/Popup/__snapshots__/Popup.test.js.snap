// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup component Popup snapshot 1`] = `
<div
  aria-hidden="true"
  className="popup popup_opened"
  onClick={[Function]}
  role="button"
  tabIndex="-1"
>
  <div
    aria-hidden="true"
    className="popup__container"
    onClick={[Function]}
    role="complementary"
    tabIndex="-1"
  >
    <button
      className="popup__close-button"
      onClick={[Function]}
      type="button"
    />
    <img
      alt="Картинка"
      className="popup__img"
      src="https://picsum.photos/id/237/600/400"
    />
    <div
      className="popup__comments-container"
    >
      <div
        className="popup__comment-container"
      >
        <p
          className="popup__comment popup__comment_type_date"
        >
          03.01.2020
        </p>
        <p
          className="popup__comment"
        >
          Крутая фотка
        </p>
      </div>
    </div>
    <form
      action="#"
      className="popup__form"
      method="get"
      noValidate={true}
      onSubmit={[Function]}
    >
      <span
        className="popup__form-error popup__form-error_hide"
      >
        
      </span>
      <input
        className="popup__form-input"
        maxLength="200"
        minLength="2"
        name="name"
        onChange={[Function]}
        pattern="^[а-яёА-ЯЁa-zA-Z0-9-\\\\s\\\\-]+$"
        placeholder="Ваше имя"
        required={true}
        type="text"
        value=""
      />
      <span
        className="popup__form-error popup__form-error_hide"
      >
        
      </span>
      <input
        className="popup__form-input"
        maxLength="200"
        minLength="2"
        name="comment"
        onChange={[Function]}
        pattern="^[а-яёА-ЯЁa-zA-Z0-9-\\\\s\\\\-]+$"
        placeholder="Ваш комментарий"
        required={true}
        type="text"
        value=""
      />
      <button
        className="popup__form-button"
        disabled={true}
        type="submit"
      >
        Оставить комментарий
      </button>
    </form>
  </div>
</div>
`;
